File: offer/ex22/KthNode.py
#!/usr/bin/env python
#-*- coding:utf-8 -*-


# -*- coding:utf-8 -*-
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

# 给定一棵二叉搜索树，请找出其中的第k小的结点。
# 例如， （5，3，7，2，4，6，8） 中，
# 按结点数值大小顺序第三小结点的值为4。
class Solution:
    def __init__(self):
        self.res = []
    # 返回对应节点TreeNode
    def KthNode(self, pRoot, k):
        
        return TreeNode(self.res[k-1])

    def inOrder(self, pRoot):
        if pRoot is None:
            return
        self.inOrder(pRoot.left)
        self.res.append(pRoot.val)
        self.inOrder(pRoot.right)

File: offer/ex22/KthNode.go
// Copyright 2019 readailib.com. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package main

import "fmt"

type TreeNode struct {
	val   int
	left  *TreeNode
	right *TreeNode
}

var (
	res = []int{}
)

// 返回第k个最小的节点
func KthNode(pRoot *TreeNode, k int) *TreeNode {
	inOrderRetraval(pRoot)
	if k < 1 && k > len(res) {
		return nil
	}

	return &TreeNode{res[k-1], nil, nil}
}

func inOrderRetraval(pRoot *TreeNode) {
	if pRoot == nil {
		return
	}

	inOrderRetraval(pRoot.left)
	res = append(res, pRoot.val)
	inOrderRetraval(pRoot.right)
}

func main() {
	pRoot := &TreeNode{
		5,
		&TreeNode{
			3,
			&TreeNode{
				2,
				nil,
				nil,
			},
			&TreeNode{
				4,
				nil,
				nil,
			},
		},
		&TreeNode{
			7,
			nil,
			nil,
		},
	}

	fmt.Println(KthNode(pRoot, 2))
}
